File: Cpp/Heapify.cpp
#include <vector>

class Solution {
public:
    /**
     * @param A: Given an integer array
     * @return: void
     */
    void solve(int pos, vector<int> &A) {
        int left , right;
        if (pos * 2 + 1 >= A.size()) left = INT_MAX;
        else left = A[pos * 2 + 1];
        if (pos * 2 + 2 >= A.size()) right = INT_MAX;
        else right = A[pos * 2 + 2];
        if (left < right && left < A[pos]) {
            A[2 * pos + 1] = A[pos];
            A[pos] = left;
            solve(2 * pos + 1, A);
        } else if (right < A[pos]) {
            A[2 * pos + 2] = A[pos];
            A[pos] = right;
            solve(2 * pos + 2, A);
        }
    }
    void heapify(vector<int> &A) {
        // write your code here
        for (int i = A.size() / 2 - 1; i >= 0; i --) {
            solve(i, A);
        }
    }
};

File: Java/Heapify.java
public class Solution {
    /**
     * @param A: Given an integer array
     * @return: void
     */
    public void helper(int[] A, int pos) {
        int left, right;
        if (pos * 2 + 1 >= A.length) left = Integer.MAX_VALUE;
        else left = A[pos * 2 + 1];
        if (pos * 2 + 2 >= A.length) right = Integer.MAX_VALUE;
        else right = A[pos * 2 + 2];
        if (left < right && A[pos] > left) {
            A[pos * 2 + 1] = A[pos];
            A[pos] = left;
            helper(A, pos * 2 + 1);
        } else if (A[pos] > right) {
            A[pos * 2 + 2] = A[pos];
            A[pos] = right;
            helper(A, pos * 2 + 2);
        }
    }
    public void heapify(int[] A) {
        // write your code here
        for (int i = A.length / 2 - 1; i >= 0; i--) {
            helper(A, i);
        }
    }
}
